src/graph/formulas/fallback.rs

```diff
@@ -0,0 +1,238 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+//! Fallback expression generator for components and meters.
+
+use crate::component_category::CategoryPredicates;
+use crate::{ComponentGraph, Edge, Error, Node};
+use std::collections::BTreeSet;
+
+use super::expr::Expr;
+
+impl<N, E> ComponentGraph<N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    /// Returns a formula expression with fallbacks where possible for the `sum`
+    /// of the given component ids.
+    pub(super) fn fallback_expr(
+        &self,
+        component_ids: impl IntoIterator<Item = u64>,
+        prefer_meters: bool,
+    ) -> Result<Expr, Error> {
+        FallbackExpr {
+            prefer_meters,
+            graph: self,
+        }
+        .generate(BTreeSet::from_iter(component_ids))
+    }
+}
+
+struct FallbackExpr<'a, N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    pub(crate) prefer_meters: bool,
+    pub(crate) graph: &'a ComponentGraph<N, E>,
+}
+
+impl<'a, N, E> FallbackExpr<'a, N, E>
+where
+    N: Node,
+    E: Edge,
+{
+    fn generate(&self, mut component_ids: BTreeSet<u64>) -> Result<Expr, Error> {
+        let mut formula = None::<Expr>;
+        while let Some(component_id) = component_ids.pop_first() {
+            if let Some(expr) = self.meter_fallback(component_id)? {
+                formula = Self::add_to_option(formula, expr);
+            } else if let Some(expr) = self.component_fallback(&mut component_ids, component_id)? {
+                formula = Self::add_to_option(formula, expr);
+            } else {
+                formula = Self::add_to_option(formula, Expr::component(component_id));
+            }
+        }
+
+        formula.ok_or(Error::internal("Search for fallback components failed."))
+    }
+
+    /// Returns a fallback expression for a meter component.
+    fn meter_fallback(&self, component_id: u64) -> Result<Option<Expr>, Error> {
+        let component = self.graph.component(component_id)?;
+        if !component.is_meter() || self.graph.has_meter_successors(component_id)? {
+            return Ok(None);
+        }
+
+        if !self.graph.has_successors(component_id)?
+            || !self
+                .graph
+                .successors(component_id)?
+                .all(|x| x.is_supported())
+        {
+            return Ok(Some(Expr::component(component_id)));
+        }
+
+        let successor_expr = self
+            .graph
+            .successors(component_id)?
+            .map(Expr::from)
+            .reduce(|a, b| a + b)
+            .ok_or(Error::internal(
+                "Can't find successors of components with successors.",
```

Comment:
`graph.successors` returns an iterator, not a set.  We can't do `len() == 0`, etc., on it, but the `has_successors` method is still O(1) because it only checks if there's at least one successor.

Because it doesn't allocate and allows us to iterate on nodes stored in the graph, this is better than making a local copy.

Reply:
Thanks a lot for explanation!